contracts/test/Proofs.t.sol

```diff
@@ -0,0 +1,158 @@
+    }
+
+    ///// Helper functions /////
+
+    function generateLeaves(uint256 count) internal pure returns (bytes32[] memory) {
+        bytes32[] memory result = new bytes32[](count);
+        for (uint256 i = 0; i < count; i++) {
+            result[i] = bytes32(i);
+        }
+        return result;
+    }
+
+    // Builds a merkle tree from an array of leaves.
+    // The tree is an array of arrays of bytes32.
+    // The last array is the leaves, and each prior array is the result of the commutative hash of pairs in the previous array.
+    // An unpaired element is simply copied to the next level.
+    // The first element of the first array is the root.
+    function buildMerkleTree(bytes32[] memory leaves) internal pure returns (bytes32[][] memory) {
+        require(leaves.length > 0, "Leaves array must not be empty");
+
+        uint256 levels = 256 - clz(leaves.length - 1);
+        bytes32[][] memory tree = new bytes32[][](levels + 1);
+        tree[levels] = leaves;
+
+        for (uint256 i = levels; i > 0; i--) {
+            bytes32[] memory currentLevel = tree[i];
+            uint256 nextLevelSize = (currentLevel.length + 1) / 2;
+            tree[i - 1] = new bytes32[](nextLevelSize);
+
+            for (uint256 j = 0; j < nextLevelSize; j++) {
+                if (2 * j + 1 < currentLevel.length) {
+                    tree[i - 1][j] = commutativeHash(currentLevel[2 * j], currentLevel[2 * j + 1]);
+                } else {
+                    tree[i - 1][j] = currentLevel[2 * j];
+                }
+            }
+        }
+
+        return tree;
+    }
+
+    // Gets an inclusion proof from a Merkle tree for a leaf at a given index.
+    // The proof is constructed by traversing up the tree to the root, and the sibling of each node is appended to the proof.
+    // There is no sibling for an unpaired element, so it is not included in the proof, which thus be shorter than the tree height.
+    function buildProof(bytes32[][] memory tree, uint256 index) internal pure returns (bytes32[] memory) {
+        require(index < tree[tree.length - 1].length, "Index out of bounds");
+
+        bytes32[] memory proof = new bytes32[](tree.length - 1);
+        uint256 proofIndex = 0;
+
+        for (uint256 i = tree.length - 1; i > 0; i--) {
+            uint256 levelSize = tree[i].length;
+            uint256 pairIndex = index ^ 1; // XOR with 1 to get the pair index
+
+            if (pairIndex < levelSize) {
+                proof[proofIndex] = tree[i][pairIndex];
+                proofIndex++;
+            }
+
+            index /= 2; // Move to the parent node
+        }
+        // Trim proof to the correct length, from proofIndex.
+        bytes32[] memory trimmedProof = new bytes32[](proofIndex);
+        for (uint256 i = 0; i < proofIndex; i++) {
+            trimmedProof[i] = proof[i];
+        }
+        return trimmedProof;
+    }
+
+    // Helper function to calculate the number of leading zeros in binary representation
```

Comment:
It turns out we're going to need this for the sum tree search function.  Could you create a library for low level binary operations and add this there?

Reply:
I'll then tweak [this function](https://github.com/FILCAT/pdp/pull/27/files#diff-712b2db21b090e745391534efb9ce9dfb70c3f729952e9433b627e264e3f83a3R212) rename to ctz, and also add to the library.